tests/tests_ut/az_ulib_ipc_ut/az_ulib_ipc_ut.c

```diff
@@ -0,0 +1,1467 @@
+  ASSERT_ARE_EQUAL(int, 0, g_count_lock);
+  ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+  /// cleanup
+  az_ulib_ipc_deinit();
+}
+
+/* If the provided handle is NULL, the az_ulib_ipc_init shall return AZ_ULIB_ILLEGAL_ARGUMENT_ERROR.
+ */
+TEST_FUNCTION(az_ulib_ipc_init_with_null_handle_failed) {
+  /// arrange
+
+  /// act
+  AZ_ULIB_RESULT result = az_ulib_ipc_init(NULL);
+
+  /// assert
+  ASSERT_ARE_EQUAL(int, AZ_ULIB_ILLEGAL_ARGUMENT_ERROR, result);
+  ASSERT_IS_NULL(g_lock);
+  ASSERT_ARE_EQUAL(int, 0, g_count_lock);
+  ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+
+  /// cleanup
+}
+
+/* If the az_ulib_ipc_init is called more then once, it shall return
+ * AZ_ULIB_ALREADY_INITIALIZED_ERROR. */
+TEST_FUNCTION(az_ulib_ipc_init_double_initialization_failed) {
+  /// arrange
+  az_ulib_ipc ipc1;
+  az_ulib_ipc ipc2;
+  ASSERT_ARE_EQUAL(int, AZ_ULIB_SUCCESS, az_ulib_ipc_init(&ipc1));
+  g_lock = NULL;
+  g_count_lock = 0;
+  umock_c_reset_all_calls();
+
+  /// act
+  AZ_ULIB_RESULT result = az_ulib_ipc_init(&ipc2);
+
+  /// assert
+  ASSERT_ARE_EQUAL(int, AZ_ULIB_ALREADY_INITIALIZED_ERROR, result);
+  ASSERT_IS_NULL(g_lock);
+  ASSERT_ARE_EQUAL(int, 0, g_count_lock);
+  ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+
+  /// cleanup
+  az_ulib_ipc_deinit();
+}
+
+/* The az_ulib_ipc_publish shall store the descriptor published in the IPC. The az_ulib_ipc_publish
+ * shall be thread safe. */
+TEST_FUNCTION(az_ulib_ipc_publish_succeed) {
+  /// arrange
+  ASSERT_ARE_EQUAL(int, az_ulib_ipc_init(&g_ipc), AZ_ULIB_SUCCESS);
+  umock_c_reset_all_calls();
+
+  STRICT_EXPECTED_CALL(az_pal_os_lock_acquire(IGNORED_PTR_ARG));
+  STRICT_EXPECTED_CALL(az_pal_os_lock_release(IGNORED_PTR_ARG));
+  STRICT_EXPECTED_CALL(az_pal_os_lock_acquire(IGNORED_PTR_ARG));
+  STRICT_EXPECTED_CALL(az_pal_os_lock_release(IGNORED_PTR_ARG));
+  STRICT_EXPECTED_CALL(az_pal_os_lock_acquire(IGNORED_PTR_ARG));
+  STRICT_EXPECTED_CALL(az_pal_os_lock_release(IGNORED_PTR_ARG));
+  STRICT_EXPECTED_CALL(az_pal_os_lock_acquire(IGNORED_PTR_ARG));
+  STRICT_EXPECTED_CALL(az_pal_os_lock_release(IGNORED_PTR_ARG));
+
+  /// act
+  ASSERT_ARE_EQUAL(int, AZ_ULIB_SUCCESS, az_ulib_ipc_publish(&MY_INTERFACE_1_123));
+  ASSERT_ARE_EQUAL(int, AZ_ULIB_SUCCESS, az_ulib_ipc_publish(&MY_INTERFACE_1_2));
+  ASSERT_ARE_EQUAL(int, AZ_ULIB_SUCCESS, az_ulib_ipc_publish(&MY_INTERFACE_2));
+  ASSERT_ARE_EQUAL(int, AZ_ULIB_SUCCESS, az_ulib_ipc_publish(&MY_INTERFACE_3));
+
+  /// assert
+  ASSERT_ARE_EQUAL(int, 0, g_count_lock);
+  ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+  az_ulib_ipc_interface_handle interface_handle;
+  ASSERT_ARE_EQUAL(
+      int,
+      AZ_ULIB_SUCCESS,
+      az_ulib_ipc_try_get_interface(
+          MY_INTERFACE_1_123.name,
+          MY_INTERFACE_1_123.version,
+          AZ_ULIB_VERSION_EQUALS_TO,
```

Comment:
>AZ_ULIB_VERSION_EQUALS_TO [](start = 10, length = 25)

not sure there are any UT for `AZ_ULIB_VERSION_LOWER_THAN` or `AZ_ULIB_VERSION_GREATER_THAN`. by design?

Reply:
No, I'll implement it.

---
In reply to: [338304880](https://github.com/Azure/azure-ulib-c/pull/40#discussion_r338304880) [](ancestors = 338304880)